documentation/src/docs/watermarking.md

```diff
@@ -0,0 +1,121 @@
+
+### Explicitly note time
+
+The following writer API method notes time.
+
+`writer.noteTime(long timestamp);`
+
+You may call the method after every event write, or use it periodically to indicate  the time below which all events are written.
+
+Here is an example:       
+
+      EventStreamWriter<EventType> writer = clientFactory.createEventWriter(stream, serializer,
+                EventWriterConfig.builder().build());
+
+      //... write events ...
+
+      writer.noteTime(currentTime);
+
+### Note time on transaction commit
+
+For transactional writes, the commit call can supply the timestamp. The following Writer API method passes the time to the Controller as part of the commit. The  commit simultaneously commits the data and supplies the corresponding timestamp.
+
+`void commit(long timestamp) throws TxnFailedException;`
+
+For example:
+
+    Transaction<EventType> txn = writer.beginTxn();
+      //... write events to transaction.
+    txn.commit(txnTimestamp);
+
+### Automatically attach wall clock time
+
+The `automaticallyNoteTime` option when a  Writer is created configures the   writer to automatically attach a wall clock time notation to every event. This option essentially attaches the ingest time to each event. With this option turned on, no other calls to note the time are required.
+
+Here is the call to configure automatic timestamps on events:
+
+    EventStreamWriter<EventType> writer = clientFactory.createEventWriter(stream, serializer,
+                EventWriterConfig.builder().automaticallyNoteTime(true).build());
+
+
+## How Pravega Readers request watermark windows  
+
+Multiple Pravega Readers working in parallel each have different current locations in the stream. The Controller coordinates the Reader positions and can return a window that indicates where in the stream all of the readers are.  
+
+The following Reader API method requests the current watermark window.
+
+`TimeWindow window = reader.getTimeWindow();`
+
+The `TimeWindow` is returned as:
+
+    @Data
+    public class TimeWindow {
+        private final long lowerTimeBound;
+        private final long upperTimeBound;
+    }
+
+where:
+* `lowerTimeBound` of the  window is the watermark. The Controller asserts that all readers are done reading all events earlier than the watermark.
+*  `upperTimeBound` is an arbitrary assignment that can help applications keep track of a moving window of time.
+
+The lower bound is most important. The lower bound is labeled the watermark. It takes into account the position of all of the readers and represents the time before which all events are "guaranteed" to have been read. Events can arrive out of order but the watermark would indicate which window each event aggregrates into. At some point, a window needs to be declared closed, so that processing can occur.  Events can also arrive late, after its window has been closed. The action on such an occurrence is up to the application.
```

Comment:
Why is guaranteed is quotes?
Also the word is not needed: "a time before which all events have been read."